tiledb/sm/stats/global_stats.cc

```diff
@@ -88,43 +88,61 @@ void GlobalStats::set_enabled(bool enabled) {
 }
 
 void GlobalStats::reset() {
-  std::unique_lock<std::mutex> ul(mtx_);
-  for (auto& register_stat : registered_stats_) {
-    register_stat->reset();
-  }
+  iterate([](Stats& stat) { stat.reset(); });
 }
 
 void GlobalStats::register_stats(const shared_ptr<Stats>& stats) {
   std::unique_lock<std::mutex> ul(mtx_);
   registered_stats_.emplace_back(stats);
 }
 
+template <class FuncT>
+void GlobalStats::iterate(const FuncT& f) {
+  std::unique_lock<std::mutex> ul(mtx_);
+  for (auto register_stat = registered_stats_.begin();
+       register_stat != registered_stats_.end();) {
+    if (std::shared_ptr<Stats> stat(register_stat->lock()); stat.use_count()) {
+      f(*stat);
+      ++register_stat;
+    } else {
+      register_stat = registered_stats_.erase(register_stat);
+    }
+  }
+}
+
+template <class FuncT>
+void GlobalStats::iterate(const FuncT& f) const {
```

Comment:
It took me a while staring at this to figure out what the function signature difference was. Deserves a comment to the effect of "const-qualified iteration does not erase list elements".